added a tutorial for bundlind a Java app with Flatpak

ryanlerch opened the pull request:
Here is a walkthrough tutorial for creating a Flatpak bundle from a pre-built Java app.

The tutorial is written pretty much to stand on it's own (other than installing flatpak), so users should be able to follow the steps and get to the end without much prior knowledge.

alexlarsson commented on source/bundling-a-java-app.html.haml.markdown:
```diff
@@ -0,0 +1,150 @@
+
+  <pre>
+  <span class="unselectable">$ </span>wget https://sdk.gnome.org/keys/gnome-sdk.gpg
+  <span class="unselectable">$ </span>flatpak remote-add --gpg-import=gnome-sdk.gpg sdk.gnome.org https://sdk.gnome.org/repo/
```
Call the repo "gnome", as that matches what the other tutorials use.

alexlarsson commented:
I think this tutorial is missing some part. It needs a desktop file with icons (with the right names "com.sweethome3d.App.desktop" / "com.sweethome3d.App.png"), and ideally an appdata file.

ryanlerch commented:
Thanks for the feedback!

I have updated the tutorial with the suggested changes.

alexlarsson commented on source/bundling-a-java-app.html.haml.markdown:
```diff
@@ -16,32 +16,32 @@
   <pre>
   <span class="unselectable">$ </span>wget https://sdk.gnome.org/keys/gnome-sdk.gpg
-  <span class="unselectable">$ </span>flatpak remote-add --gpg-import=gnome-sdk.gpg sdk.gnome.org https://sdk.gnome.org/repo/
+  <span class="unselectable">$ </span>flatpak --user remote-add --gpg-import=gnome-sdk.gpg gnome https://sdk.gnome.org/repo/
```
Oh, you don't need --user here anymore. These days we have policykit authenticated system-wide remotes. 
Just use --user for the things you locally build (like your app).

alexlarsson commented on source/bundling-a-java-app.html.haml.markdown:
```diff
@@ -16,32 +16,32 @@
-  <span class="unselectable">$ </span>flatpak install gnome org.freedesktop.Platform 1.4
-  <span class="unselectable">$ </span>flatpak install gnome org.freedesktop.Sdk 1.4
+  <span class="unselectable">$ </span>flatpak --user install gnome org.freedesktop.Platform 1.4
+  <span class="unselectable">$ </span>flatpak --user install gnome org.freedesktop.Sdk 1.4
```
Same here, install these without --user to make them system-wide installed.

alexlarsson commented on source/bundling-a-java-app.html.haml.markdown:
```diff
@@ -62,14 +70,94 @@
+  <pre>
+  <span class="unselectable">$ </span>flatpak build sweethome3d-flatpak mkdir -p /app/share/appdata/
+  <span class="unselectable">$ </span>flatpak build sweethome3d-flatpak cp com.sweethome3d.App.appdata.xml /app/share/appdata/
+  </pre>
```
This is a great start, although its missing a last step, which admittedly is not obvious or well documented.
After installing the desktop file and the appdata file you need to run something like
```
flatpak build sweethome3d-flatpak appstream-compose --prefix=/app --origin=flatpak --basename=com.sweethome3d.App  com.sweethome3d.App
```
This will read the desktop file, the icons and the appdata file and combine these details into an appstream files in `/app/share/app-info/`. These files are what flatpak extracts from every app in a repo and puts into the appstream branch.

For an example you can do:
```
flatpak update --appstream gnome-nightly-apps
```
Then look into `/var/lib/flatpak/appstream/gnome-nightly-apps/x86_64/active` to see how this looks.
This is the data that gnome-software uses for a remote.

alexlarsson replied:
Oh, note that the appstream-compose stuff is done automatically with flatpak-builder.

ryanlerch commented:
Removed the extra refereneces to the --user flag, and added an extra section about the generation of appstream files, and noted that flatpak-builder does this automatically.

jimmac commented:
Does this look good to go, Alex?

alexlarsson commented:
Looks pretty good to me. One issue is that the flatpak-builder version doesn't add the desktop file and the appdata file. We can add that later though.
